native/dlasq2.go

```diff
@@ -0,0 +1,330 @@
+// successfully. If info is 1, a split was marked by a positive value in e.
+// If info is 2, current block of Z not diagonalized after 100*n
+// iterations (in inner while loop).  On exit Z holds a qd array with the same
+// eigenvalues as the given Z. If info is 3,  termination criterion of outer
+// while loop not met (program created more than N unreduced blocks).
+func (impl Implementation) Dlasq2(n int, z []float64) (info int) {
+	// TODO(btracey): make info an error.
+	if len(z) < 4*n {
+		panic("lapack: insufficient z length")
+	}
+	const cbias = 1.5
+
+	eps := dlamchP
+	safmin := dlamchS
+	tol := eps * 100
+	tol2 := tol * tol
+	if n < 0 {
+		panic(nLT0)
+	}
+	if n == 0 {
+		return info
+	}
+	if n == 1 {
+		if z[0] < 0 {
+			panic("lapack: negative z value")
+		}
+		return info
+	}
+	if n == 2 {
+		if z[1] < 0 || z[2] < 0 {
+			panic("lapack: bad z value")
+		} else if z[2] > z[0] {
+			z[0], z[2] = z[2], z[0]
+		}
+		z[4] = z[0] + z[1] + z[2]
+		if z[1] > z[2]*tol2 {
+			t := 0.5 * (z[0] - z[2] + z[1])
+			s := z[2] * (z[1] / t)
+			if s <= t {
+				s = z[2] * (z[1] / (t * (1 + math.Sqrt(1+s/t))))
+			} else {
+				s = z[2] * (z[1] / (t + math.Sqrt(t)*math.Sqrt(t+s)))
+			}
+			t = z[0] + s + z[1]
+			z[2] *= z[0] / t
+			z[0] = t
+		}
+		z[1] = z[2]
+		z[5] = z[1] + z[0]
+		return info
+	}
+	z[2*n-1] = 0
+	emin := z[1]
+	var d, e, qmax, zmax float64
+	var i1, n1 int
+	for k := 0; k < 2*(n-1); k += 2 {
+		if z[k] < 0 {
+			panic("lapack: bad z value")
+		} else if z[k+1] < 0 {
+			panic("lapack: bad z value")
+		}
+		d += z[k]
+		e += z[k+1]
+		qmax = math.Max(qmax, z[k])
+		emin = math.Min(emin, z[k+1])
+		zmax = math.Max(math.Max(qmax, zmax), z[k+1])
+	}
+	if z[2*(n-1)] < 0 {
+		panic("lapack: bad z value")
+	}
+	d += z[2*(n-1)]
+	qmax = math.Max(qmax, z[2*(n-1)])
+	zmax = math.Max(qmax, zmax)
+	if e == 0 {
+		for k := 1; k < n; k++ {
+			z[k] = z[2*k]
+		}
+		impl.Dlasrt(lapack.SortDecreasing, n, z)
+		z[2*(n-1)] = d
+		return info
+	}
+	trace := d + e
+	if trace == 0 {
+		z[2*(n-1)] = 0
+		return info
+	}
+
+	for k := 2 * n; k >= 2; k -= 2 {
+		z[2*k-1] = 0
+		z[2*k-2] = z[k-1]
+		z[2*k-3] = 0
+		z[2*k-4] = z[k-2]
+	}
+	i0 := 0
+	n0 := n - 1
+
+	if cbias*z[4*i0] < z[4*n0] {
+		ipn4Out := 4 * (i0 + n0 + 2)
+		for i4loop := 4 * (i0 + 1); i4loop <= 2*(i0+n0+1); i4loop += 4 {
+			i4 := i4loop - 1
+			ipn4 := ipn4Out - 1
+			z[i4-3], z[ipn4-i4-4] = z[ipn4-i4-4], z[i4-3]
+			z[i4-1], z[ipn4-i4-6] = z[ipn4-i4-6], z[i4-1]
+		}
+	}
+
+	pp := 0
+	for k := 0; k < 2; k++ {
+		d = z[4*n0+pp]
+		for i4loop := 4*n0 + pp; i4loop >= 4*(i0+1)+pp; i4loop -= 4 {
+			i4 := i4loop - 1
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
+				d = z[i4-3]
+			} else {
+				d = z[i4-3] * (d / (d + z[i4-1]))
+			}
+		}
+		emin = z[4*(i0+1)+pp]
+		d = z[4*i0+pp]
+		for i4loop := 4*(i0+1) + pp; i4loop <= 4*n0+pp; i4loop += 4 {
+			i4 := i4loop - 1
+			z[i4-2*pp-2] = d + z[i4-1]
+			if z[i4-1] <= tol2*d {
+				z[i4-1] = math.Copysign(0, -1)
```

Comment:
I tried with setting the element to 1 (and +Inf) and it passes too - so I don't know how valuable a pass with just changing the sign of zero is.

The only place where it would make a difference is when you take the max and get the index (which you can't unless you explicitly test for the sign bit, which doesn't happen here - at least in dlasq2, maybe in the subroutines it does).